# scVI Integration with HCA Lung Reference

Jointly trains scVI on in-house data (fused, parental, resistant) and a public HCA lung reference dataset. 
The batch covariate is `source` (in_house vs public_HCA), which lets scVI remove technical variation 
between the two origins while placing your cells in the context of known lung cell types.

**Workflow:**
1. Load in-house 10x data
2. Download HCA lung data via CellxGene Census
3. Verify raw counts in HCA data
4. Concatenate with `join='inner'` (shared genes only)
5. QC filter
6. HVG selection with `batch_key='source'`
7. scVI with `batch_key='source'`, `categorical_covariate_keys=['sample']`
8. Embedding, clustering, visualization
9. DE between in-house sample groups (restricted to in-house cells)

## Imports

In [1]:
import numpy as np
import pandas as pd
import scanpy as sc
import anndata as ad
import scvi
import cellxgene_census  # pip install cellxgene-census

sc.set_figure_params(figsize=(5, 5))
%config InlineBackend.print_figure_kwargs={'facecolor': 'w'}
%config InlineBackend.figure_format='retina'

## 1. Load In-House Data

In [2]:
adata_fus = sc.read_10x_mtx('../fused/', var_names='gene_symbols', cache=True)
adata_fus.obs['sample'] = 'fused'
adata_fus.obs.index = [f'F_{bc}' for bc in adata_fus.obs.index]

adata_par = sc.read_10x_mtx('../parental/', var_names='gene_symbols', cache=True)
adata_par.obs['sample'] = 'parental'
adata_par.obs.index = [f'P_{bc}' for bc in adata_par.obs.index]

adata_res = sc.read_10x_mtx('../resistant/', var_names='gene_symbols', cache=True)
adata_res.obs['sample'] = 'resistant'
adata_res.obs.index = [f'R_{bc}' for bc in adata_res.obs.index]

for adata in [adata_fus, adata_par, adata_res]:
    adata.var_names_make_unique()
    adata.obs['source'] = 'in_house'

print(adata_fus)
print(adata_par)
print(adata_res)

AnnData object with n_obs × n_vars = 6044 × 36601
    obs: 'sample', 'source'
    var: 'gene_ids', 'feature_types'
AnnData object with n_obs × n_vars = 10353 × 36601
    obs: 'sample', 'source'
    var: 'gene_ids', 'feature_types'
AnnData object with n_obs × n_vars = 8513 × 36601
    obs: 'sample', 'source'
    var: 'gene_ids', 'feature_types'


## 2. Download HCA Lung Reference via CellxGene Census

Streams lung cells directly — no need to download the full atlas.
Set `census_version` to a pinned release for reproducibility; check available versions at
https://chanzuckerberg.github.io/cellxgene-census/cellxgene_census_docsite_data_release_info.html

In [ ]:
#import ssl
#import urllib.request

## SSL workaround for macOS certificate issues
#ssl_context = ssl.create_default_context()
#ssl_context.check_hostname = False
#ssl_context.verify_mode = ssl.CERT_NONE

#ssl._create_default_https_context = ssl._create_unverified_context

CENSUS_VERSION = '2025-11-08'  # pin for reproducibility

census = cellxgene_census.open_soma(census_version=CENSUS_VERSION)

adata_hca = cellxgene_census.get_anndata(
    census,
    organism='Homo sapiens',
    obs_value_filter="tissue_general == 'lung' and is_primary_data == True",
)

census.close()

print(adata_hca)
print('\nHCA obs columns:', adata_hca.obs.columns.tolist())

## 3. Verify HCA Raw Counts

scVI requires raw integer counts. CellxGene Census stores raw counts in `.X` by default, but confirm here.

In [ ]:
import scipy.sparse as sp

sample_vals = adata_hca.X[:10, :10]
if sp.issparse(sample_vals):
    sample_vals = sample_vals.toarray()

print('Sample HCA X values (should be non-negative integers):')
print(sample_vals)
print('\nAll integer:', np.all(sample_vals == np.floor(sample_vals)))

In [ ]:
adata_hca.obs['sample'] = 'HCA_lung'
adata_hca.obs['source'] = 'public_HCA'

# Make barcodes unique to avoid collisions with in-house barcodes
adata_hca.obs.index = [f'HCA_{bc}' for bc in adata_hca.obs.index]

print(adata_hca)

## 4. Concatenate — Inner Join on Shared Genes

`join='inner'` keeps only genes present in both the in-house data and HCA.
Your in-house data has 36,601 genes; the HCA Census uses Ensembl gene IDs mapped to symbols,
so the intersection will be somewhat smaller.

In [ ]:
# Concatenate in-house samples first (same gene space), then add HCA
adata_inhouse = ad.concat([adata_fus, adata_par, adata_res], join='outer')

adata_combined = ad.concat(
    [adata_inhouse, adata_hca],
    join='inner',   # keeps only genes shared between in-house and HCA
    merge='same',
)

print(adata_combined)
print(f'\nShared genes: {adata_combined.n_vars}')
print(f'Source breakdown:')
print(adata_combined.obs['source'].value_counts())

## 5. QC Filtering

In [ ]:
sc.pp.filter_cells(adata_combined, min_genes=200)
sc.pp.filter_genes(adata_combined, min_cells=3)

adata_combined.var['mt'] = adata_combined.var_names.str.startswith('MT-')
sc.pp.calculate_qc_metrics(
    adata_combined, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True
)

# Visualize QC metrics split by source before filtering
sc.pl.violin(
    adata_combined,
    ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'],
    groupby='source',
    jitter=0.4,
    multi_panel=True,
)

In [ ]:
adata_combined_filtered = adata_combined[adata_combined.obs.total_counts < 75000, :]
adata_combined_filtered = adata_combined_filtered[adata_combined_filtered.obs.n_genes > 1000, :]
adata_combined_filtered = adata_combined_filtered[adata_combined_filtered.obs.pct_counts_mt < 40, :]

print(adata_combined_filtered)
print(adata_combined_filtered.obs['source'].value_counts())

## 6. Normalize, Log-Transform, and Select HVGs

HVGs are selected with `batch_key='source'` so genes are ranked by variability within each source 
independently, then combined. This avoids selecting genes that vary only because of source differences.

In [ ]:
adata_combined_filtered = adata_combined_filtered.copy()  # materialize view
adata_combined_filtered.layers['counts'] = adata_combined_filtered.X.copy()

sc.pp.normalize_total(adata_combined_filtered, target_sum=1e6)
sc.pp.log1p(adata_combined_filtered)
adata_combined_filtered.raw = adata_combined_filtered

sc.pp.highly_variable_genes(
    adata_combined_filtered,
    n_top_genes=3000,        # increased from 1200 — more genes improve integration
    subset=True,
    layer='counts',
    flavor='seurat_v3',
    batch_key='source',      # rank HVGs within each source independently
)

print(f'HVGs selected: {adata_combined_filtered.n_vars}')

## 7. scVI Setup and Training

- `batch_key='source'`: corrects for in-house vs public technical differences  
- `categorical_covariate_keys=['sample']`: also accounts for fused/parental/resistant variation

In [ ]:
scvi.model.SCVI.setup_anndata(
    adata_combined_filtered,
    layer='counts',
    batch_key='source',
    categorical_covariate_keys=['sample'],
)

In [ ]:
model = scvi.model.SCVI(adata_combined_filtered, n_latent=30)
model.train(accelerator='mps')  # use GPU (Apple Silicon)

In [ ]:
# Save model so you don't have to retrain
model.save('scvi_HCA_integration_model/', overwrite=True)

## 8. Latent Representation and Clustering

In [ ]:
SCVI_LATENT_KEY = 'X_scVI'
adata_combined_filtered.obsm[SCVI_LATENT_KEY] = model.get_latent_representation()

sc.pp.neighbors(adata_combined_filtered, use_rep=SCVI_LATENT_KEY)

SCVI_CLUSTERS_KEY = 'leiden_scVI'
sc.tl.leiden(
    adata_combined_filtered,
    key_added=SCVI_CLUSTERS_KEY,
    resolution=0.5,
    flavor='igraph',
    n_iterations=2,
    directed=False,
)

sc.tl.umap(adata_combined_filtered, min_dist=0.3)

## 9. Visualization

Key plots to assess integration quality and cell type context.

In [ ]:
# Integration quality: in-house and HCA cells should intermingle if integration worked
sc.pl.umap(
    adata_combined_filtered,
    color=['source'],
    frameon=False,
    title='Source (integration check)',
)

In [ ]:
# In-house sample identity
sc.pl.umap(
    adata_combined_filtered,
    color=['sample'],
    frameon=False,
    groups=['fused', 'parental', 'resistant'],  # exclude HCA_lung from color
    title='In-house samples',
)

In [ ]:
# Leiden clusters
sc.pl.umap(
    adata_combined_filtered,
    color=[SCVI_CLUSTERS_KEY],
    frameon=False,
    legend_loc='on data',
)

In [ ]:
# HCA cell type annotations — shows which lung cell types your cells resemble
# 'cell_type' is the standard column in CellxGene Census objects
if 'cell_type' in adata_combined_filtered.obs.columns:
    sc.pl.umap(
        adata_combined_filtered,
        color=['cell_type'],
        frameon=False,
        legend_fontsize=7,
        title='Cell type (HCA annotations)',
    )
else:
    print('No cell_type column found. Check adata_combined_filtered.obs.columns')

## 10. Differential Expression — In-House Samples Only

Restrict DE to in-house cells so we compare fused/parental/resistant against each other,
not against HCA lung cells.

In [ ]:
adata_inhouse_filtered = adata_combined_filtered[
    adata_combined_filtered.obs['source'] == 'in_house'
].copy()

print(adata_inhouse_filtered)
print(adata_inhouse_filtered.obs['sample'].value_counts())

In [ ]:
# Re-load the model and get latent for in-house subset only
# (model.differential_expression can subset internally via indices)
de_df = model.differential_expression(
    adata=adata_inhouse_filtered,
    groupby='sample',
    mode='change',
)
de_df.head()

In [ ]:
markers = {}
cats = {'fused', 'resistant', 'parental'}
for c in cats:
    cid = f'{c} vs Rest'
    cell_type_df = de_df.loc[de_df.comparison == cid]
    cell_type_df = cell_type_df[cell_type_df.lfc_mean > 0]
    cell_type_df = cell_type_df[cell_type_df['bayes_factor'] > 3]
    cell_type_df = cell_type_df[cell_type_df['non_zeros_proportion1'] > 0.1]
    markers[c] = cell_type_df.index.tolist()[:3]

print('Top markers per sample:', markers)

In [ ]:
sc.tl.dendrogram(adata_inhouse_filtered, groupby='sample', use_rep='X_scVI')

sc.pl.dotplot(
    adata_inhouse_filtered,
    markers,
    groupby='sample',
    dendrogram=True,
    color_map='Blues',
    swap_axes=True,
    use_raw=True,
    standard_scale='var',
)

## 11. Cell Type Composition of In-House Samples

Shows what fraction of each in-house sample maps to each HCA-annotated cluster.

In [ ]:
if 'cell_type' in adata_inhouse_filtered.obs.columns:
    composition = (
        adata_inhouse_filtered.obs
        .groupby(['sample', 'cell_type'])
        .size()
        .unstack(fill_value=0)
    )
    composition_norm = composition.div(composition.sum(axis=1), axis=0)

    composition_norm.T.plot(
        kind='bar',
        stacked=False,
        figsize=(14, 5),
        title='Cell type composition by in-house sample (HCA annotations)',
    )
    import matplotlib.pyplot as plt
    plt.ylabel('Fraction of cells')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    plt.show()
else:
    print('No cell_type column — run cell type annotation first.')

## 12. UMAP — In-House Cells Only

Plot just in-house cells on the shared latent UMAP to see where fused/parental/resistant land 
relative to the HCA-defined clusters.

In [ ]:
sc.pl.umap(
    adata_inhouse_filtered,
    color=['sample', SCVI_CLUSTERS_KEY],
    frameon=False,
    ncols=2,
)